mava/utils/checkpointing.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+import typing
+import warnings
+from copy import copy
+from datetime import datetime
+from pydoc import locate
+from typing import Any, Dict, Optional, Union
+
+import absl.logging as absl_logging
+import orbax.checkpoint
+from chex import Numeric
+from flax.core.frozen_dict import FrozenDict
+from jax.tree_util import tree_map
+from omegaconf import DictConfig
+
+from mava.types import HiddenStates, LearnerState, Params, RNNLearnerState
+
+CHECKPOINTER_VERSION = 0.1
+
+
+class Checkpointer:
+    """Model checkpointer for saving and restoring the `learner_state`."""
+
+    def __init__(
+        self,
+        model_name: str,
+        metadata: Optional[Dict] = None,
+        rel_dir: str = "checkpoints",
+        timestamp_override: Optional[str] = None,
+        save_interval_steps: int = 1,
+        max_to_keep: Optional[int] = 1,
+        keep_period: Optional[int] = None,
+    ):
+        """Initialise the checkpointer tool
+
+        Args:
+            model_name (str): Name of the model to be saved.
+            metadata (Optional[Dict], optional):
+                For storing model metadata. Defaults to None.
+            rel_dir (str, optional):
+                Relative directory of checkpoints. Defaults to "checkpoints".
+            timestamp_override (Optional[str], optional):
+                Set the dir name within rel_dir/model_name/<...>.
+                Defaults to None, which means the timestamp is used.
+            save_interval_steps (int, optional):
+                The interval at which checkpoints should be saved. Defaults to 1.
+            max_to_keep (Optional[int], optional):
+                Maximum number of checkpoints to keep. Defaults to 1.
+            keep_period (Optional[int], optional):
+                If set, will not delete any checkpoint where
+                checkpoint_step % keep_period == 0. Defaults to None.
+        """
+
+        # Sharding info will be read from file, rather than from 'RestoreArgs'
+        warnings.filterwarnings(
+            action="ignore",
+            category=UserWarning,
+            message="Couldn't find sharding info under RestoreArgs",
+        )
```

Comment:
Just a question, what is this and why do we suppress this warning?

Reply:
Originally I was getting an unnecessary warning thrown about how I was saving a sharded array, but it's no longer there—so I can actually remove it! 🖖🏻

Reply:
If we hit it again in the future, we can revisit :)

Reply:
Sounds good, thanks man :)

Reply:
Oh wait, sorry! The warning is thrown when loading checkpoints. The full error message is
```
UserWarning: Couldn't find sharding info under RestoreArgs. Populating sharding info from sharding file. Please note restoration time will be slightly increased due to reading from file instead of directly from RestoreArgs.
```
So it uses the checkpointed file for sharding info, which is what we want (I believe), rather than providing it in the code. I think we can happily suppress the warning?

Reply:
Sounds good. Thanks :)